[launch] Change launch shortcuts to use facade pattern

kriegfrj opened the pull request:
This makes it easier to interactively test them during development.

There should be no functional change from a user perspective as a result of this update (possible exception: I may have fixed a bug in the "native" launcher, which to be honest I've never used or understood its purpose). However, I would like to do some work soon to update the launching, and having this fix in-place first will make it much easier to do that work.

pkriens commented:
I like what you're doing with the services. However, I think we're creating another blob if we aggregate by the fact that something is  service. BJ explained that you can leverage this decomposition in testing. That is very nice but I do not think this should drive decomposition. 

So I would highly recommend to create a quickfix & launch bundle. I am more than ok to just have the services there but over time we can maybe grow them to have all the associated functionality.

After working with many codebases where decompositions were done for these reasons I've become allergic to this :-) It invariably create the next blob with too many dependencies. The primary value of modularity is that it is an iceberg. A small part above the water but the far majority below. Using the wrong decomposition destroys that idea. It is interesting to read the [seminal paper from Parnas](https://prl.ccs.neu.edu/img/p-tr-1971.pdf) to see how he defined modularity. Basically he tried to decompose a problem in different ways and found that some decompositions were _much_ more maintainable. Key concept is _cohesion_.

So for now the syncing I want in core until my quest to get the workspace model is good enough. It is too close to Central. For the quickfix and launch shortcuts I'd suggest to create separate bundles named by their function.

kriegfrj commented:
Thanks for the suggestion.

The reason I put it in the services bundle was because I could restart the bundle while Eclipse was running. This significantly reduces the time in the develop- test cycle because you don't have to wait for Eclipse to restart every time you make a small change. 

Rather than have lots of little bundles, I'd like in the long term to see two - one that contains only those things that can't restart, and one that contains everything else. I don't think there's a massive need to decompose it further.

Admittedly, in the in between state where it is at the moment, it can be a bit awkward with dependencies split across the bundles. The launch shortcuts also had this problem. I'm still playing with it to try and figure out the sweet spot. 

In the meantime, I understand if you want to move it back, but for your own sanity delay this so you can take advantage of the restart ability in your own development.

kriegfrj commented:
Sorry, I think I misunderstood. You're in favor of decomposing into multiple smaller service bundles. I can do that. I was avoiding it because it would be a bigger change, architecturally speaking.

pkriens commented:
I think it is a great idea but if we do such an effort, my experience tells me that 'temporarily' aggregating the services in a services bundle will be quite permanent. If we make any effort, we should at least move in the correct _direction_ . Which is imho to provide a cohesive place for each function. I.e. a launch (services) bundle and a quickfix (services) bundle. Over time we can then move more and more of the closely related parts there as well.

kriegfrj commented:
> my experience tells me that 'temporarily' aggregating the services in a services bundle will be quite permanent. 

😆 It is also my experience that many a temporary situation becomes quasi-permanent.

I'm just pausing on this PR for a while, as over the weekend I discovered that Eclipse supposedly has the concept of a dynamic plugin: https://wiki.eclipse.org/FAQ_What_is_a_dynamic_plug-in%3F The Workbench plugin is already dynamic aware (so the above page claims). Thus it should in theory be possible to re-install `bndtools.core` and have things keep working. This raises the question of why we get those "registry object invalid" messages when we restart `bndtools.core`. It's possible we're simply not cleaning up after ourselves properly somewhere.

In the case of the QuickFix processor, I recall that the list of active processors are cached internally by the JDT plugin in a static variable, so this forces the use of a facade to achieve dynamicity. But maybe with other chunks of functionality, this would not be necessary.

Regardless of the answer to this question, modularising Bndtools is likely to be a useful thing to do. So perhaps I'll put the launch code in its own bundle, make sure it follows fully dynamic practices, and see if it works *without* a facade. That would certainly simplify things.

I also came up with a dynamic proxy implementation of the facade pattern, which means you don't have to write custom facades for every extension point (as long as the extension point is specified by an interface rather than a class - ie, most extension points). If a facade is necessary, perhaps I will change this PR to use the dynamic proxy approach.

juergen-albert commented:
I like the idea of the face pattern. I support Peters veto to the temporary solution at the same time. In Regard to the Dynamic Plugin s you have mentioned: According to the JavaDoc of the mentioned shutdown Method it is already deprecated and the `BundleActivator.stop` method should be used instead. As this is around for ages now, I would not count on better support for a dynamic world anytime soon.

Therefor I would suggest, that we create one facade bundle, where we move all extensionpoints from bndtools.core and that provides service Interfaces the core can implement and provide.

pkriens commented:
For my understanding, remove plugin.xml from core and move it to anther bundle, allowing core to implement services?

Sounds like a big bang. My intuition tells me it is better to carve cohesive pieces out of the core and do them as good as we can, using what we know now. 

But I am open to be convinced.

If we're moving in this direction, we might want to create an API project in the workspace that holds **all** service interfaces. I find that that model works very well for many different development dimensions.

juergen-albert commented:
True, doing this in one big move will be too much of a bang. My actual point is, what Fr. Krieg is already doing. separate the Extensionpoints from the actual logic. I agree, that we should do this step by step.

For the API bundle: Makes sense. 
On how we divide the logic and how many bundles we end up with that hold extension points: I'm open for everything. We just should come up with a pattern we agree on. I would opt for the the one bundle with all the extensions (we can move them there step by step). Just to know where to find them and to have a starting point. For the rest, we can have as many bundles as will make sense.

pkriens commented:
As I argued initially, a bundle with extension points sounds like the wrong composition. Maybe we should have a call where Fr. Krieg can also attend? I'd like to know a bit more about this.

juergen-albert commented:
+1

kriegfrj commented:
If I may summarise the discussion so far, there are two separate (but related) goals:

1. Implementing facade pattern
2. Modularising bndtools.core

As you guys noted, we like to try and avoid big bangs (especially when we don't have good test coverage, which is a bit of a chicken-and-egg problem at the moment), so I've been focusing on 1 to keep it simple. I also wasn't sure about how the rest of the team might feel about the explosion of bundles that might arise out of 2. But the feeling I'm getting is that we don't mind an explosion of bundles (as long as they are coherent). Early experience has also taught me that it will probably be more difficult to focus on 1 without looking at 2 at the same time. So the way forward seems to be:

1. Modularise (one subsystem at a time, so that we avoid a big bang).
2. Implement the facade (where appropriate) on each of the newly-created subsystem modules.

Does that sound fair to you guys?

I also think we need to clarify something: we're talking about "bundles that hold the extension points". In the facade pattern, there are two ends - the facade (that registers the extension in the Eclipse extension registry and interfaces to the OSGi service registry), and the actual implementation. The idea is that you put as much of the code as you can in the implementation bundles and the least possible in the facade to keep Eclipse happy at runtime. I'd be happy with one bundle to register all of the facades (which should be fairly small), but individual implementation bundles. Otherwise we'd double the size of the "explosion".

As for an API bundle: I think we already have one? `bndtools.api`? Also, for Eclipse plugin implementations, the API interface is usually defined in an Eclipse bundle somewhere, so for such services we already effectively have an API bundle too (eg, the quick fix processor component is in a private package and implements `IQuickFixProcessor` from the JDT bundle).

Looking down the track, once we bed down the facade pattern to something that works well in a variety of circumstances, I'm dreaming of tooling to automatically generate some or all of the facade itself (eg, perhaps generate Req-Cap directives on the facade and Provide-Cap on the implementation bundles so that the resolver can assemble it all reliably).

I'm happy to try and jump on a Bndtools call at some point to discuss this if you'd like (I just can't make the bndtools calls regularly due to the time).

kriegfrj commented:
Btw, I also stumbled upon this which @njbartlett wrote, which in spite of its age is relevant to the current discussion:

https://www.slideshare.net/mfrancis/extensions-vs-services-digging-deeper-neil-bartlett

In summary, extensions are in theory dynamic, but most of the clients of extension have not been written dynamic aware which  means that in practice they are not. The facade pattern, then, is not so much a solution for the inherent non-dynamicity of the extension registry, but a way to insulate non-dynamic-aware code from the dynamicity of components that they are using. In theory, a poorly written OSGi service consumer could also be non-dynamic-aware, and the facade pattern would be equally useful in such a circumstance.

pkriens commented:
> As for an API bundle: I think we already have one? bndtools.api? 

That is an ok place but I wan to be sure that bundles can **only** communicate through interfaces defined in one project. This project is fine to me. I think we also need a bnd.api bundle probably.

A lot of things become simpler if you have API classes are separate.

We have a call next friday, could you joing Fr. Krieg? I think we could dedicate it to this topic.

kriegfrj commented:
Here's a bit of a snapshot of the latest effort.

* I have moved the launch functionality into its own bundle.
* There's a generic utility called `ExtensionFacade` that can be used in place of pretty much any executable extension whose API is an interface. You only need to tell it what the base interface is that it out to be emulating.
* You can successfully stop/start the launch bundle without having to restart your running Eclipse. You currently get an "unhandled event loop exception" if you try and use any of the launch bundle's functionality while it's not running.
* I'm experimenting with being able to enable/disable the corresponding GUI elements based on whether or not the backing service for the extension's facade

kriegfrj commented:
Here is the latest effort.

I think in its current form, apart from the need to tidy up/squash the commit history and the DCO, this is ready to merge. But before I go to the effort of a major cleanup, I'd appreciate it if someone could spare some cycles for a high-level review pass.

Notable changes in this latest push (noting some of this for my own future reference as well as anyone else who might be curious):
* I made some changes to the facade to make sure that it shuts down the tracker when it is finalized.
* TrackerCustomizer now uses a weak reference back to the facade rather than a strong reference, so that it doesn't prevent the facade from being garbage collected (and the tracker from being automatically closed).
* Where necessary (eg, the instantiated extension objects are stateful), I changed some of the extension services to use PROTOTYPE scope. The facade's tracker customizer has been updated to handle this transparently.
* I discovered that there is some client code in the launch manager that isn't content with checking to see if the extension class implements the appropriate interface, but looks for an appropriate abstract base class. Eg [here](https://git.eclipse.org/c/platform/eclipse.platform.debug.git/tree/org.eclipse.debug.ui/ui/org/eclipse/debug/internal/ui/sourcelookup/SourceLookupPanel.java?id=7076decdceeadf55f1dc9aa6a3bb65ecf5b64f83#n177), where it insists on an instance of `AbstractSourceLookupDirector`. The synthetic proxy that implemented `ISourceLookupDirector` did not pass the `instanceof` test (even though the object it was proxying does implement `AbstractSourceLookupDirector`), and so the source lookup info failed to show in the GUI.
 The workaround for this was to introduce a new mode where the synthetic proxy is bypassed and a direct reference to the service object instance itself (which passes the `instanceof` test) is returned. The disadvantage of this mode is that we relinquish control over the lifecycle of the object and so we can't insulate the caller from the launch service bundle restarting. However, in this case the lifecycle of the component is short and under control through the UI - closing the relevant UI tab releases the reference to the object anyway, so in practice it doesn't stop you restarting the bundle. This mode wouldn't be suitable for extension components that are cached internally for a "long time" (eg, the quick fix processor which is stored in a static variable).

kriegfrj commented:
Some open questions I have for any interested reviewers:
* Best way to log (what I've put in there now was a rough implementation to help me with debugging).
* Best way to report when the backing service is missing in the proxy `invoke()` and `create()` methods.

kriegfrj commented:
Any takers for this one?

A couple more notes: I hacked in a couple of package exports because of dependencies that `org.bndtools.launch` needed. I think I can probably get rid of those and I can remove the exports.

gamerson commented:
I will try to take this for a spin sometime next week and give some
feedback.  I've wanted a few improvements around the launches (and
debugging source lookup) so I can try to give some constructive feedback.

On Mon, Jul 26, 2021 at 10:35 AM Fr Jeremy Krieg ***@***.***>
wrote:

> Any takers for this one?
>
> A couple more notes: I hacked in a couple of package exports because of
> dependencies that org.bndtools.launch needed. I think I can probably get
> rid of those and I can remove the exports.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bndtools/bnd/pull/4703#issuecomment-886809025>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAERKFLNAPTIA37KWMRI54DTZV6EPANCNFSM46NDYV5Q>
> .
>


-- 
Greg Amerson
Liferay Developer Tools
Liferay, Inc. www.liferay.com

kriegfrj commented:
> A couple more notes: I hacked in a couple of package exports because of dependencies that `org.bndtools.launch` needed. I think I can probably get rid of those and I can remove the exports.

In the latest push, I managed to get rid of most of these package exports. The only one remaining is for the package `bndtools`, and that is because the launch bundle needs access to `bndtools.Plugin` for logging and also `bndtools.StatusCode`. On some level it seemed to make sense to make this package accessible but I think that could be revisited and refactored too.

I also managed to get rid of the gratuitous class references in `bndtools.Plugin` that were necessary to force Bnd to generate imports for those classes. Now, the `ExtensionFacade` uses the defining bundle of the extension point to load the extension class/interface, which should usually work as the bundle defining the extension point will normally also contain the interfaces that the extension point uses.

Still interested on people's ideas about the best way to handle the error cases, and if anyone wants to take it for a spin...

gamerson commented:
Hey @kriegfrj  I rebased this PR onto master, fired it up on the `bndtools.test/bndws` project root.  When I right-clicked `bndtools.test/bndws/launch.simple/really.bndrun` and did `Run As > Bnd OSGi Run Launcher` I got the following errors in the log:

```
java.lang.RuntimeException: Service bndtools.launch.RunShortcut (org.eclipse.debug.ui.ILaunchShortcut2) not found
	at org.bndtools.facade.ExtensionFacade.lambda$1(ExtensionFacade.java:206)
	at java.util.Optional.orElseThrow(Optional.java:290)
	at org.bndtools.facade.ExtensionFacade.getRequiredService(ExtensionFacade.java:199)
	at org.bndtools.facade.ExtensionFacade.invoke(ExtensionFacade.java:275)
	at com.sun.proxy.$Proxy52.launch(Unknown Source)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchShortcutExtension.launch(LaunchShortcutExtension.java:430)
	at org.eclipse.debug.internal.ui.actions.LaunchShortcutAction.runInternal(LaunchShortcutAction.java:88)
	at org.eclipse.debug.internal.ui.actions.LaunchShortcutAction.runWithEvent(LaunchShortcutAction.java:135)
```

```
java.lang.RuntimeException: Service bndtools.launch.RunShortcut (org.eclipse.debug.ui.ILaunchShortcut2) not found
	at org.bndtools.facade.ExtensionFacade.lambda$1(ExtensionFacade.java:206)
	at java.util.Optional.orElseThrow(Optional.java:290)
	at org.bndtools.facade.ExtensionFacade.getRequiredService(ExtensionFacade.java:199)
	at org.bndtools.facade.ExtensionFacade.invoke(ExtensionFacade.java:275)
	at com.sun.proxy.$Proxy52.launch(Unknown Source)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchShortcutExtension.launch(LaunchShortcutExtension.java:430)
	at org.eclipse.debug.internal.ui.actions.LaunchShortcutAction.runInternal(LaunchShortcutAction.java:88)
```
```
null
bndtools.core
Warning
Wed Aug 18 10:21:56 CDT 2021
Service bndtools.launch.RunShortcut (org.eclipse.debug.ui.ILaunchShortcut2) not found.
```

kriegfrj commented:
> Hey @kriegfrj I rebased this PR onto master, fired it up on the `bndtools.test/bndws` project root. When I right-clicked `bndtools.test/bndws/launch.simple/really.bndrun` and did `Run As > Bnd OSGi Run Launcher` I got the following errors in the log:
> [snip]
> null
> bndtools.core
> Warning
> Wed Aug 18 10:21:56 CDT 2021
> Service bndtools.launch.RunShortcut (org.eclipse.debug.ui.ILaunchShortcut2) not found.
> ```

These are the errors that you get when the backing service is not running. You should also get some helpful diagnostics in the OSGi console for the test Eclipse instance.

It seems that although I re-resolved the win32 bndrun, I did not re-resolve the MacOS or GTK bndruns. So the new bundle `org.bndtools.launch` was not added to these launches. Hence the backing services were missing.

I have fixed this, rebased on master, and re-pushed. If you would be so kind as to take it for another spin, that would be much appreciated. Although I also discovered through this process that the `bndws` workspace is in a bit of a state of disrepair at the moment too and doesn't seem to function well out-of-the-box, with broken paths to libraries, missing bundles, etc...

bjhargrave commented on bndtools.api/bnd.bnd:
```diff
@@ -10,12 +10,16 @@
 
 Import-Package: \
  ${eclipse.importpackage},\
+ org.eclipse.debug.ui,\
```
This seems wrong. The macro should import the proper packages.

kriegfrj replied:
Good catch; fixed.

bjhargrave commented on bndtools.api/src/org/bndtools/api/editor/package-info.java:
```diff
@@ -0,0 +1,3 @@
+@org.osgi.annotation.bundle.Export
```
We should not use `@Export` in this repo. Export from the bnd file.

kriegfrj replied:
This was a cut-and-paste from `org.bndtools.api`. I have fixed all instances in the bundle, including `org.bndtools.api`.

bjhargrave reviewed (approved):
LGTM

kriegfrj commented:
> LGTM

👍 Are we going to merge it then, or are we still waiting for @gamerson to take it for a spin?

gamerson commented:
> > LGTM
> 
> 👍 Are we going to merge it then, or are we still waiting for @gamerson to take it for a spin?

gamerson commented:
Looks good to me.

kriegfrj commented:
Changed all the `System.err.println()` statements to use SLF4J at an appropriate level instead.
